### PR TITLE
ICU-20627 Adding getAvailableLocalesByType, ICU4C and ICU4J.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20627
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/unicode/uloc.h [hunk 1]
```diff
@@ -742,12 +742,19 @@ uloc_getDisplayName(const char* localeID,
...
+ *
+ * This method corresponds to uloc_openAvailableByType called with the
+ * ULOC_AVAILABLE_DEFAULT type argument.
+ *
+ * The return value is a pointer to an item of a locale name array. Both this
+ * array and the pointers it contains are owned by ICU and should not be
+ * deleted or written through by the caller. The locale name is terminated by
+ * a null pointer.
+ *
+ * @param flags Type choice from ULocAvailableType.
```
Comment (markusicu) [inline icu4c/source/common/unicode/uloc.h]: there are no flags

File: icu4c/source/common/unicode/uloc.h [hunk 2]
```diff
@@ -762,6 +769,70 @@ uloc_getAvailable(int32_t n);
...
+   * ULOC_AVAILABLE_WITH_LEGACY_ALIASES.
+   *
+   * @draft ICU 65
+   */
+  ULOC_AVAILABLE_ONLY_LEGACY_ALIASES,
+
+  /**
+   * The union of the locales in ULOC_AVAILABLE_DEFAULT and
+   * ULOC_AVAILABLE_ONLY_LEGACY_ALIAS.
+   */
```
Comment (markusicu) [inline icu4c/source/common/unicode/uloc.h]: @draft ICU 65

File: icu4c/source/common/unicode/uloc.h [hunk 3]
```diff
@@ -762,6 +769,70 @@ uloc_getAvailable(int32_t n);
...
+  ULOC_AVAILABLE_COUNT
+#endif
+} ULocAvailableType;
+
+/**
+ * Gets a list of available locales according to the type argument, allowing
+ * the user to access different sets of supported locales in ICU.
+ *
+ * The returned UEnumeration must be free'd by the caller.
```
Comment (markusicu) [inline icu4c/source/common/unicode/uloc.h]: not "free'd" -- "closed" as in uenum_close()

File: icu4c/source/common/locavailable.cpp [hunk 4]
```diff
@@ -95,85 +97,170 @@ U_NAMESPACE_USE
...
+  public:
+    void put(const char *key, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) U_OVERRIDE {
+        ResourceTable resIndexTable = value.getTable(status);
+        if (U_FAILURE(status)) {
+            return;
+        }
+        for (int32_t i = 0; resIndexTable.getKeyAndValue(i, key, value); ++i) {
+            int32_t type;
+            if (uprv_strcmp(key, "InstalledLocales") == 0) {
+                type = 0;
```
Comment (markusicu) [inline icu4c/source/common/locavailable.cpp]: In Java you are using the enum constants. Why not here?

File: icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java [hunk 5]
```diff
@@ -786,11 +836,37 @@ public int compareTo(ULocale other) {
...
+     * Returns a list of all installed locales according to the specified type.
+     *
+     * @draft ICU 65
+     * @provisional This API might change or be removed in a future release.
+     */
+    public static Collection<ULocale> getAvailableLocalesByType(AvailableType type) {
+        if (type == null) {
+            throw new IllegalArgumentException();
+        }
+        if (type == ULocale.AvailableType.WITH_LEGACY_ALIASES) {
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java]: Optional: Before this "if", declare a List<ULocale> list. Then in each branch set the list, and after the else body, have a single shared `return Collections.unmodifiableList(list);`.

File: icu4c/source/common/locavailable.cpp (hunk 4 above)
Comment (sffc) [reply icu4c/source/common/locavailable.cpp]: I was, until I changed the enum to having a third type, in which case the fields no longer correspond to each element in the enum.  I was considering changing Java but liked the convenience of the EnumMap.

I added an assertion to Java.

File: icu4c/source/common/locavailable.cpp [hunk 6]
```diff
@@ -95,85 +97,170 @@ U_NAMESPACE_USE
...
+            } else {
+                // CLDRVersion, etc.
+                continue;
+            }
+            ResourceTable availableLocalesTable = value.getTable(status);
+            if (U_FAILURE(status)) {
+                return;
+            }
+            gAvailableLocaleCounts[type] = availableLocalesTable.getSize();
+            gAvailableLocaleNames[type] = static_cast<const char**>(
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/common/locavailable.cpp]: Done

File: icu4c/source/common/locavailable.cpp [hunk 7]
```diff
@@ -95,85 +97,170 @@ U_NAMESPACE_USE
...
 }
 
 U_CAPI int32_t  U_EXPORT2
-uloc_countAvailable()
-{
-    _load_installedLocales();
-    return _installedLocalesCount;
+uloc_countAvailable() {
+    icu::ErrorCode status;
+    _load_installedLocales(*status);
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/common/locavailable.cpp]: Done

File: icu4c/source/common/locavailable.cpp [hunk 8]
```diff
@@ -95,85 +97,170 @@ U_NAMESPACE_USE
...
+    if (U_FAILURE(*status)) {
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/common/locavailable.cpp]: Done

File: icu4c/source/common/unicode/uloc.h (hunk 1 above)
Comment (sffc) [reply icu4c/source/common/unicode/uloc.h]: Done

File: icu4c/source/common/unicode/uloc.h (hunk 2 above)
Comment (sffc) [reply icu4c/source/common/unicode/uloc.h]: Done.

File: icu4c/source/common/unicode/uloc.h (hunk 3 above)
Comment (sffc) [reply icu4c/source/common/unicode/uloc.h]: Done.

File: icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java (hunk 5 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java]: Done.

File: icu4c/source/common/locavailable.cpp (hunk 4 above)
Comment (sffc) [reply icu4c/source/common/locavailable.cpp]: Done as requested.

**Pull request merged**